### ISSUE TITLE
build instructions on EL8

### ISSUE THREAD

**Issue by markusdd** [NVIDIA/enroot#130]
To make it simple: Despite following the instructions what packages are required etc the 'elaborate' ;) build instructions of 'make install' do not work on my EL8 system (up to date Rocky Linux 8.6).

As there are no EL8 rpms in the release process I am forced to build it myself, but to no avail.

I get this message:
![image](https://user-images.githubusercontent.com/25175069/181102213-e422d4fa-bda4-46b0-849d-354d5c7e9131.png)

I tried to understand what Makefile is trying to do there, but I do not get it. Isn't it supposed to use the musl distribution it clones?

Is there any more elaborate build documentation?

I am stuck...although I have actually quite a bit of experience custom-compiling software. But this does not follow any standard configure or cmake-style buildflow and I wasn't able to find any further docs on the Makefile system sued in this project.

Any advice?

P.S.: In an effort to 'reverse-engineer' the Makefile I tried setting the FORCE_GLIBC env variable, this runs some build process, but ultimately gets stuck here: 
![image](https://user-images.githubusercontent.com/25175069/181102882-b5227628-c70e-47f5-b292-17f408ab1b44.png)

This system has been used to build all kind of software for the cluster, so the gcc, g++ etc. build tool are all there and most definitely work.

**Comment by 3XX0** [comment]
The RHEL packages we ship should work on any distribution, everything is statically linked.

**Comment by markusdd** [comment]
I can confirm they don't on EL8. At least for me.
I get complete hang-up when trying to start a container. Also, on the release page they are marked explicitly as EL7.

But anyway, the build process should work for anyone. But I made some progress meanwhile. When using a fresh RockyLinux 8 container and installing the dependency everything compiles. My guess is the Makefile cannot cope with the presence of ccache in the PATH.

I get to the point where the RPMS do build, but rpmlint throws all of this as errors:
![image](https://user-images.githubusercontent.com/25175069/181121608-932b217d-dbfa-4e14-8fb0-4f52744b4b26.png)

**Comment by markusdd** [comment]
Unfortunately, same result with the self-built rpms. Upon starting a container, I end up with this.
I tried this on a clean new RockyLinux 8 VM and on the target server where we wish to use it, same problem.
Frustrating.
![image](https://user-images.githubusercontent.com/25175069/181125523-6dd54aff-3db0-4cb0-ba94-eb1778965623.png)

**Comment by flx42** [comment]
This might be a kernel bug, I think I've faced similar issues on Ubuntu 22.04 with fuse-overlayfs.

 I suggest to start by trying without relying on fuse-overlayfs, that is to say use `enroot import ; enroot create ; enroot start` instead of just `enroot import ; enroot start image.sqsh` as described in https://github.com/NVIDIA/enroot/blob/master/doc/cmd/start.md#starting-container-images

**Comment by markusdd** [comment]
Thanks for the hint, I will try.

**Comment by markusdd** [comment]
wow, ok, that worked.

So my question is now: we want to roll this out on a productive system. Is there any way (configuration or build time) to prevent usage of fuse-overlayfs?
Because this creates hanging processes which can only be removed by a reboot, very bad situation for a shared compute system.

Of course, it would actually even better if this would work, but if you say Ubuntu 22.04 is affected this alleged bug would affect a super wide range of kernels. (EL8 uses 4.18, Ubuntu 22.04 is certainly using 5.14 or newer).

**Comment by flx42** [comment]
> So my question is now: we want to roll this out on a productive system. Is there any way (configuration or build time) to prevent usage of fuse-overlayfs?

Simply: do not install `fuse-overlayfs` and `squashfuse` on your system. Of course a user could still install those 2 applications from sources and point at those binaries with a custom $PATH, but that seems like an unlikely workflow.

I confirm that I'm still seeing issues on Ubuntu 22.04 with the 5.17 OEM kernel, and SIGKILL does not work.

**Comment by flx42** [comment]
Also you could try unloading `fuse` if it's a kernel module on your distro (looks like it's builtin on Ubuntu 22.04).

**Comment by markusdd** [comment]
I guess removing squashfuse would be an intermediate solution.
Fuse itself is required for other puposes. But thanks for your hints, that is helpful as a workaround for now.

**Comment by flx42** [comment]
BTW I was told by a Canonical support engineer that you can add `initcall_blacklist=fuse_init` to the kernel boot parameters to prevent FUSE from loading even if it is builtin. In case you don't need it anymore of if someone else looks at this issue in the future.

**Comment by markusdd** [comment]
Ok, so RedHat came back to me and the analyed the issue and claim it is an issue by enroot:

I quote from the ticket:

"that is a bug in enroot.

I see the current fuse-overlayfs process running:

fuse-overlayfs -f -o lowerdir=/run/user/1000/enroot/overlay/lower,upperdir=/run/user/1000/enroot/overlay/upper,workdir=/run/user/1000/enroot/overlay/work /run/user/1000/enroot/overlay

That means the overlay mount is created at /run/user/1000/enroot/overlay but the lower layer is also under that directory.  So when fuse-overlayfs needs to do a lookup on the lower directory (/run/user/1000/enroot/overlay/lower), will end up accessing its own mount (/run/user/1000/enroot/overlay).

Please report the issue upstream to enroot, since it is not a supported RHEL package."

**Comment by flx42** [comment]
Maybe so, but if enroot (an unprivileged process) can put a system in a bad state, then it is a risk for system stability and it probably needs to be fixed in the Linux kernel (unless I'm missing something).

**Comment by markusdd** [comment]
I made that  statement as well, but the system is not really 'in a bad state', everything else runs perfectly fine.
It is just due to the 'circle dependency' the 2 fuse processes hang themselves up and frankly the kernel has no real chance of knowing how to resolve that. The requests being made are valid, but due to mounting that stuff in the way enroot does, it creates a classic deadlock situation.

Probably fuse would need to account for that by disallowing that.

Nevertheless, I think the point they are making that enroot is doing the mount requests in an incorrect manner is valid.
So even if fuse would'nt hang, it would need to decline whatever enroot tries to do.

Does that make sense?

So yes, Fuse could most likely improve here, nevertheless enroot has a problem here I'd say.

**Comment by flx42** [comment]
> I made that statement as well, but the system is not really 'in a bad state', everything else runs perfectly fine.

Can you kill the fuse-overlayfs with SIGTERM or SIGKILL? If it's unkillable, I would argue the system is indeed in a bad state.

**Comment by flx42** [comment]
@3XX0 would have to take a look at the FUSE operations done by enroot when he's back.

**Comment by markusdd** [comment]
> > I made that statement as well, but the system is not really 'in a bad state', everything else runs perfectly fine.
> 
> Can you kill the fuse-overlayfs with SIGTERM or SIGKILL? If it's unkillable, I would argue the system is indeed in a bad state.

No, it cannot be. You can free up the tty again by killing the bash process that launched enroot.
But the 2 fuse processes that wait for each other are stuck forever waiting for each other resources.

So no, it does not affect anything else on the server, but these 2 are stuck until reboot. So it is not an unstable, but dirty state.

**Comment by flx42** [comment]
Right, but if you use Slurm, it will not like this dirty state and will drain the node:
https://slurm.schedmd.com/slurm.conf.html#OPT_UnkillableStepTimeout

And I think this behavior makes sense, it usually indicates something might be very wrong with the system.

**Comment by markusdd** [comment]
True, but for SLURM it will not really make a difference as the jobs would also fail if fuse would behave more elegant, as the circular mount dependency requested by enroot would make the job crash as well.

The only way currently is to uninstall squashfuse to make it impossible to run enroot start immidiately on a squashfs image.
This will do for now, but fundamentally the way enroot uses fuse_overlayfs must be fixed I guess.

**Comment by 3XX0** [comment]
Yeah I will look at it when I have time, but IIRC this shouldn't be a problem and I fixed several issues in fuse-overlayfs in the past that made it work. So I guess there has been some changes there.

In any case on newer kernels we should transition to unprivileged overlay and get rid of fuse-overlayfs

**Comment by flx42** [comment]
BTW I found two ways to cleanup this situation, the first one requires root and requires going into each mount namespace that is currently stuck:
```console
sudo nsenter -t $(pidof fuse-overlayfs) -m umount -f -l /run/enroot/user-11838/overlay
```

The second one can be done unprivileged and does not require handling mount namespaces:
```console
echo 1 > /sys/fs/fuse/connections/<ID>/abort
```
To find the right value of `<ID>`, a cleanup script would have to look for folders in `/sys/fs/fuse/connections/` that are owned by the user:
```
$ ls -l /sys/fs/fuse/connections/
total 0
dr-x------ 2 root       root       0 Jul 27 12:15 44
dr-x------ 2 fabecassis fabecassis 0 Jul 27 13:34 50
```

**Comment by flx42** [comment]
> Yeah I will look at it when I have time, but IIRC this shouldn't be a problem and I fixed several issues in fuse-overlayfs in the past that made it work. So I guess there has been some changes there.

You are absolutely right, it worked fine in all versions from 0.7 (in Ubuntu 20.04) to 1.6 (one version before what is shipped in Ubuntu 22.04. I filed a bug to hopefully clarify this situation once and for all: https://github.com/containers/fuse-overlayfs/issues/364
